core/src/Kokkos_TypeInfo.hpp

```diff
@@ -0,0 +1,92 @@
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 4.0
+//       Copyright (2022) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_TYPE_INFO_HPP
+#define KOKKOS_TYPE_INFO_HPP
+
+#include <array>
+#include <string_view>
+#include <utility>
+
+#include <Kokkos_Macros.hpp>
+
+#ifndef KOKKOS_COMPILER_INTEL
+
+namespace Kokkos {
+namespace Impl {
+
+template <std::size_t... Is>
+constexpr auto string_view_to_char_array(std::string_view s,
+                                         std::index_sequence<Is...>) {
+  return std::array{s[Is]...};
+}
+
+template <class T>
+constexpr auto type_name() {
+#if defined(__clang__)
+  constexpr std::string_view func = __PRETTY_FUNCTION__;
+  constexpr std::string_view prefix{"[T = "};
+  constexpr std::string_view suffix{"]"};
+#elif defined(__GNUC__)
+  constexpr std::string_view func = __PRETTY_FUNCTION__;
+  constexpr std::string_view prefix{"[with T = "};
+  constexpr std::string_view suffix{"]"};
+#elif defined(_MSC_VER)
+  constexpr std::string_view func = __FUNCSIG__;
+  constexpr std::string_view prefix{"type_name<"};
+  constexpr std::string_view suffix{">(void)"};
+#else
+#error bug
+#endif
+  constexpr auto beg = func.find(prefix) + prefix.size();
+  constexpr auto end = func.rfind(suffix);
+  static_assert(beg != std::string_view::npos);
+  static_assert(end != std::string_view::npos);
+  static_assert(beg < end);
+  constexpr auto name = func.substr(beg, end - beg);
+  return string_view_to_char_array(name,
+                                   std::make_index_sequence<name.size()>());
+}
+
+}  // namespace Impl
+
+template <class T>
+class TypeInfo {
+  static constexpr auto value_ = Impl::type_name<T>();
+
+ public:
+  static constexpr std::string_view name() noexcept {
+    return {value_.data(), value_.size()};
+  }
+};
+
+}  // namespace Kokkos
+
+#else  // out of luck, using Intel C++ Compiler Classic
+
+namespace Kokkos {
+
+template <class T>
+class TypeInfo {
+ public:
+  static constexpr std::string_view name() noexcept { return "not supported"; }
```

Comment:
I would have returned here: `typeid(T).name();`, I get it that means its not constexpr but do we really need this string has a constexpr thing? I.e. does Kokkos::TypeInfo need to be compile time where ISO type_info is not? This would also isolate the ifdefing to just this one place not everywhere where it's used.

Reply:
It would not be usable in a constant expression in that case (that is the solution you suggest would not have the `constexpr` which means already a departure from the non ICC code path).